### PR TITLE
feat: add project name normalization with friendly name aliasing and metadata tracking

### PR DESCRIPTION
**Pull request by kevinpjones**
## Summary

- **Project name normalization**: Clients can now use human-friendly project names (with spaces, special characters, unicode) that are automatically normalized to filesystem-safe directory names via a new `normalizeProjectName()` utility
- **Metadata & indexing system**: Each project gets a `.metadata.json` file tracking `friendlyName`, `directoryName`, and `createdAt`; a global `.index` file provides efficient friendly-name-to-directory-name lookup with file-locking for concurrent safety
- **MCP adapter-layer resolution**: A `withProjectResolution` higher-order wrapper at the adapter layer transparently resolves friendly names for all MCP tools, with lazy metadata backfill for existing projects on first access
- **Enriched API responses**: `list_projects` now returns objects with both `name` (directory) and `friendlyName` fields
- **Web UI updates**: Project list and detail pages display friendly names as primary titles with directory names as subtitles

## Architecture

The implementation follows the existing Clean Architecture patterns:

| Layer | New Components |
|-------|---------------|
| **Domain** | `ProjectMetadata`, `ProjectInfo` entity types |
| **Data** | `normalizeProjectName()` helper, `ProjectNameResolverService`, `rebuildProjectIndex` utility, `MetadataRepository` / `ProjectIndexRepository` protocols |
| **Infra** | `FsMetadataRepository`, `FsProjectIndexRepository` (with `FileLockService`) |
| **Main** | Factory functions, `mcp-project-resolution-adapter.ts` wrapper, updated `routes.ts` |
| **Web UI** | Updated `MemoryBankService`, `ProjectList`, project detail page |

Key design decision: Resolution is handled at the **MCP adapter layer** via a higher-order function, minimizing changes to existing controllers, use cases, and repositories. This keeps the cross-cutting concern centralized and easily extensible for future features (e.g., project tags).

## Test plan

- [x] 643 tests passing across 46 test files
- [x] 100% coverage on all new data/infra/adapter code (6 new test files, 108 new tests)
- [x] TypeScript compilation passes (`npm run build`)
- [x] All existing tests continue to pass (backwards compatible)
- [x] Lazy metadata backfill tested for existing projects without `.metadata.json`
- [x] Collision detection tested for normalized name conflicts
- [x] Unicode transliteration and edge cases covered in normalization tests

Made with [Cursor](https://cursor.com)

---

> [!NOTE]
> **Medium Risk**
> Adds new project name resolution and persistence paths (metadata files + global index) and changes MCP tool inputs/outputs (e.g., `list_projects`) while introducing background index rebuild on startup, so regressions could affect project routing and filesystem writes.
> 
> **Overview**
> **Adds friendly project names end-to-end** by introducing `normalizeProjectName`, a `ProjectNameResolver`, and a new `.metadata.json` + global `.index` mapping layer (with locking) to map user-friendly names to filesystem-safe directory names.
> 
> **Updates the MCP layer** to transparently resolve `projectName` for all read/write tools via `withProjectResolution`, backfill missing metadata on first access, and enrich `list_projects` responses to return `{ name, friendlyName }` objects; also triggers an async `rebuildProjectIndex` on MCP server startup.
> 
> **Polishes filesystem/UI behavior** by hiding dotfiles (e.g., `.metadata.json`, `.index`) from `listFiles`, and updating the web UI + service layer to display/search by `friendlyName` while still using directory `name` for routing.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit da6b39339393595f889576309620576deaf8a242. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/main/protocols/mcp/adapters/mcp-project-resolution-adapter.ts [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+      friendlyName: directoryName,
+      directoryName,
+      createdAt: new Date().toISOString(),
+    };
+    return metadataRepo
+      .writeMetadata(directoryName, metadata)
+      .then(() => indexRepo.setMapping(directoryName, directoryName))
+      .catch(() => {}); // Swallow errors — lazy backfill is best-effort
+  }).catch(() => {}); // Swallow errors
+}
```
[the remark this answers is not in the document]
Comment (kevinpjones) [reply src/main/protocols/mcp/adapters/mcp-project-resolution-adapter.ts]: Resolved by [5f328ac](https://github.com/kevinpjones/memory-bank-mcp/pull/17/commits/5f328ac6a772788be36322a76ba2cd66028592cc)

File: src/main/protocols/mcp/routes.ts [hunk 2]
```diff
@@ -419,4 +485,64 @@ export default () => {
...
+
+  /**
+   * Creates an enriched list_projects handler that returns ProjectInfo objects
+   * (with both directory name and friendly name) instead of plain strings.
+   */
+  function createEnrichedListProjectsHandler(): Promise<(request: any) => Promise<any>> {
+    const baseHandler = adaptMcpRequestHandler(makeListProjectsController());
+
+    return baseHandler.then((handler) => {
+      return async (request: any) => {
```
[the remark this answers is not in the document]
Comment (kevinpjones) [reply src/main/protocols/mcp/routes.ts]: Resolved by [0a88908](https://github.com/kevinpjones/memory-bank-mcp/pull/17/commits/0a88908fb30064fa46606d2a99fe614175f48a42)

File: src/data/helpers/normalize-project-name.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    throw new Error("Project name cannot be empty");
+  }
+
+  let normalized = input.trim().toLowerCase();
+
+  // Transliterate common Unicode characters
+  normalized = normalized
+    .split("")
+    .map((char) => TRANSLITERATIONS[char] ?? char)
+    .join("");
```
[the remark this answers is not in the document]
Comment (kevinpjones) [reply src/data/helpers/normalize-project-name.ts]: Resolved by [4fa9913](https://github.com/kevinpjones/memory-bank-mcp/pull/17/commits/4fa9913bbc7b172fe76ee208342f180317421e2c)

File: src/main/protocols/mcp/routes.ts [hunk 4]
```diff
@@ -15,14 +15,53 @@ import {
...
+
+/**
+ * Shape of a tool response from the MCP request adapter.
+ * The SDK's ServerResult is a union type, so we define the shape we expect.
+ */
+interface ToolResponse {
+  content?: Array<{ type: string; text: string }>;
+  isError?: boolean;
+  [key: string]: unknown;
+}
```
[the remark this answers is not in the document]
Comment (kevinpjones) [reply src/main/protocols/mcp/routes.ts]: Resolved by [22d3951](https://github.com/kevinpjones/memory-bank-mcp/pull/17/commits/22d3951c1197b63e74bdc653f3e704fd1515aba7)
